ramips: add support for Arcadyan WG630223

system1357 opened the pull request:
Arcadyan WG630223 is a wireless AX1800 (WiFi 6) mesh router. It uses
the same case design as MTS WG430223.

Device specification
--------------------
SoC Type: MediaTek MT7621AT
RAM: 256 MiB
Flash: 128 MiB (Toshiba TC58BVG0S3HTA00 SLC SPI NAND)
Wireless 2.4 GHz (MT7915DA): b/g/n/ax, 2x2
Wireless 5 GHz (MT7915DN): a/n/ac/ax, 2x2
Ethernet: 3xGbE (WAN, LAN1, LAN2)
USB ports: No
Button: 1 (Reset/WPS)
LEDs: 3 (Red, Blue, Green)
Power: 12 VDC, 1 A
Connector type: Barrel
Bootloader: U-Boot (Version: 2018.09_TC-0.2)
Also called: CHT Wi-Fi Home-NFP (Wi-Fi 6_2T2R)

Installation(ssh)
------------
1. Login to the router web interface (username: cht, password: 23tc +
    last 4 digits of label mac address)
2. Navigate to System -> Remote Management -> enable ssh for LAN
3. Connect to router using scp (credentials same as web interface)
    to upload firmware, default folder is /tmp/cht_home
4. Connect to the router using ssh (credentials same as web interface)
5. Flash OpenWrt image:
```
ls -l (this is to find out file length)
mtd_debug erase /dev/mtd4 0 25165824
mtd_debug write /dev/mtd4 0 "file length" "file name"
```
9. Reboot into OpenWrt

Installation(TFTP / uboot console)
------------

1. Prepare a TFTP server (tested with tftpd32)
2. Connect TX and GND pins first (UART provision is near the reset button)
    Pin arrangement from arrow sign to J4 silkscreen is PWR TX RX GND
    Do not ever connect power pin, don't connect RX just yet
3. Serial console baud rate 115200 8N1, connect RX once there's output.
    Press any key to stop auto boot, beware that the window is tight
4. (optional) Set bootdelay to longer period for easier uboot access
`env set bootdelay 10`
5. Flash OpenWrt image:
```
run 'mtkautoboot'
> Press arrow down key immediately to stop autoboot
> Select option "Upgrade firmware"
> Select upgrade method "tftp", xmodem is slow but works as well
> Fill in network data and update file name, press enter
```
6. Router auto boots into OpenWrt

Stock firmware image format
---------------------------
The same as MTS WG630223 but with another trx magic "HDR0"

Others
---------------------------
This branch also contains the necessary fix for decrypting the
board_config partition using different keys

csharper2005 commented on target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh:
```diff
@@ -51,6 +51,9 @@ platform_do_upgrade() {
 	ampedwireless,ally-r1900k|\
+	arcadyan,wg630223)
+		nand_do_upgrade "$1"
+		;;
```
This will brake other devices. You need only one line here.
`arcadyan,wg630223|\`

csharper2005 commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -9,6 +9,9 @@ ramips_setup_interfaces()
 	ampedwireless,ally-00x19k|\
+	arcadyan,wg630223)
+		ucidef_set_interfaces_lan_wan "lan1 lan2" "wan"
+		;;
```
This also brakes other devices.

csharper2005 commented on target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts:
```diff
@@ -0,0 +1,175 @@
+	};
+};
+
+&state_default {
```
Do we need to remove this block after https://github.com/openwrt/openwrt/pull/10238?

system1357 replied:
I've removed it for now

csharper2005 commented on target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts:
```diff
@@ -0,0 +1,175 @@
+
+		key-restart {
+			label = "reset";
+			gpios = <&gpio 476 GPIO_ACTIVE_LOW>;
```
Hmm. Really 476?

system1357 replied:
Ooops, that's indeed wrong

jospezial commented:
If it has ssh, does the stock firmware have sftp server or client? sshfs?
A pipe over ssh would also be possible.
Then nobody needs to setup a http server.

system1357 commented:
> If it has ssh, does the stock firmware have sftp server or client? sshfs? A pipe over ssh would also be possible. Then nobody needs to setup a http server.

Fixed PR info, added tftp/uboot and ssh flash instructions

csharper2005 commented:
@system1357, In general looks good to me. I would suggest:
1. Make "u-boot-env" writeable;
2. Add MAC address examples to the commit message.

And I would move the changes to base-files and already supported devices in a separate commit, e.g. "base-files: Allow custom encryption key for Arcadyan".

Perhaps maintainers have a different opinion on this matter. Let's wait for their review.

jospezial commented:
> * don't connect RX just yet
> 
> * Serial console baud rate 115200 8N1, connect RX once there's output.

Why? What happens if it is connected before?

system1357 commented:
> Why? What happens if it is connected before?

There will be no serial output. But if the TX pin is connected after the serial console is active(started printing something), everything works fine throughout the whole powerup session
(yes, even after a reboot command)

I've tested with a few different serial console program and adapters with different chips, all seems to show the same behavior

system1357 commented:
> @system1357 Please take a look at this PR #10451

I believe this device has a DFS antenna, which the stock firmware also utilize for scanning
(The one at lower right)
![6411D2F1-7457-4F64-826E-A462FDC42F17](https://user-images.githubusercontent.com/7140960/188851526-21d33e9b-67ac-4168-a499-7e3ec8e1aaf0.png)

DragonBluep commented on target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts:
```diff
@@ -0,0 +1,175 @@
+
+&nand {
+	status = "okay";
+	mediatek,nmbm;
```
It's better to add `mediatek,bmt-remap-range`. NMBM should exclude ubi partition. refer to [actiontec web7200](https://git.openwrt.org/?p=openwrt/staging/nbd.git;a=blob;f=target/linux/ramips/dts/mt7621_actiontec_web7200.dts)

system1357 replied:
May I ask:
If the second and third firmware partition uses squashfs/jffs2 layout, shoud I include them in the remap range?

DragonBluep replied:
yes, it should be `mediatek,bmt-remap-range = <0x0000000 0x05c0000>, <0x1980000 0x3800000>;`.  Maybe we need to use u-boot to restore the original firmware.

DragonBluep commented on target/linux/ramips/dts/mt7621_arcadyan_wg630223.dts:
```diff
@@ -0,0 +1,175 @@
+			label = "firmware_mp";
+			reg = <0x3180000 0x1800000>;
+		};
+		partition@4f80000 {
```
It is recommended to add a new line above.

DragonBluep commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -171,6 +171,27 @@ define Device/ampedwireless_ally-00x19k
 
+define Device/arcadyan_wg630223
+  $(Device/dsa-migration)
+  $(Device/uimage-lzma-loader)
```
The new u-boot 2018.09 should no longer need the lzma loader.

DragonBluep commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -171,6 +171,27 @@ define Device/ampedwireless_ally-00x19k
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | \
+	uImage none | arcadyan-trx 0x30524448 | pad-to $$(KERNEL_SIZE)
```
u-boot 2018.09 supports fit image. Maybe you can try it if you have time. 0f068e7c4a83bcbf20c4e52a5f8a3f1fe2af2246 13308161788c98ae6cd48c22b13339fdb8c77130

system1357 replied:
I'm actually running a FIT image on my own device, will push fixes later

system1357 committed:
base-files: Fix Arcadyan router mac fetch function

The original code has a hardcoded key
Changed to accept different encryption keys found in different models

Signed-off-by: Lenaius Chang <system1357@gmail.com>

system1357 committed:
ramips: add support for Arcadyan WG630223

Arcadyan WG630223 is a wireless AX1800 (WiFi 6) mesh router. It uses
the same case design as MTS WG430223.

Device specification
--------------------
SoC Type: MediaTek MT7621AT
RAM: 256 MiB
Flash: 128 MiB (Toshiba TC58BVG0S3HTA00 SLC SPI NAND)
Wireless 2.4 GHz (MT7915DA): b/g/n/ax, 2x2
Wireless 5 GHz (MT7915DN): a/n/ac/ax, 2x2
Ethernet: 3xGbE (WAN, LAN1, LAN2)
USB ports: No
Button: 1 (Reset/WPS)
LEDs: 2 (Red, Blue)
Power: 12 VDC, 1 A
Connector type: Barrel
Bootloader: U-Boot (Version: 2018.09_TC-0.2)
Also called: CHT Wi-Fi Home-NFP (Wi-Fi 6_2T2R)

Installation(ssh)
------------
1. Login to the router web interface (username: cht, password: 23tc +
    last 4 digits of label mac address)
2. Navigate to System -> Remote Management -> enable ssh for LAN
3. Connect to router using scp (credentials same as web interface)
    to upload firmware, default folder is /tmp/cht_home
4. Connect to the router using ssh (credentials same as web interface)
5. Flash OpenWrt image:

   > ls -l (this is to find out file length)
   > mtd_debug erase /dev/mtd4 0 25165824
   > mtd_debug write /dev/mtd4 0 "file length" "file name"

9. Reboot into OpenWrt

Installation(TFTP / uboot console)
------------

1. Prepare a TFTP server (tested with tftpd32)
2. Connect TX and GND pins first (UART provision is near the reset button)
    Pin arrangement from arrow sign to J4 silkscreen is PWR TX RX GND
    Do not ever connect power pin, don't connect RX just yet
3. Serial console baud rate 115200 8N1, connect RX once there's output.
    Press any key to stop auto boot, beware that the window is tight
4. (optional) Set bootdelay to longer period for easier uboot access

   > env set bootdelay 10`

5. Flash OpenWrt image:

   > run 'mtkautoboot'
   > Press arrow down key immediately to stop autoboot
   > Select option "Upgrade firmware"
   > Select upgrade method "tftp", xmodem is slow but works as well
   > Fill in network data and update file name, press enter

6. Router auto boots into OpenWrt

MAC addresses
-------------
+-----------+-------------------+
| Interface | MAC               |
+-----------+-------------------+
| label     | A8:xx:xx:C0:xx:14 |
| LAN       | A8:xx:xx:C0:xx:14 |
| WAN       | A8:xx:xx:C0:xx:17 |
| WLAN_2g   | AA:xx:xx:50:xx:14 |
| WLAN_5g   | AA:xx:xx:70:xx:14 |
+-----------+-------------------+

Stock firmware image format
---------------------------
The same as MTS WG630223 but with another trx magic "HDR0"

Signed-off-by: Lenaius Chang <system1357@gmail.com>

system1357 committed:
Merge branch 'openwrt:master' into wg630223

DragonBluep commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -190,6 +190,26 @@ define Device/ampedwireless_ally-00x19k
+  KERNEL_SIZE := 4352k
+  UBINIZE_OPTS := -E 5
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
```
Please replace these nand common definitions with `$(Device/nand)`.
ref: https://github.com/openwrt/openwrt/commit/b09a83808a3a4f6ff2ff0cf1f12fdc4193357896
